bftengine/src/bftengine/ReadOnlyReplica.cpp

```diff
@@ -101,19 +102,27 @@ void ReadOnlyReplica::onMessage<CheckpointMsg>(CheckpointMsg *msg) {
   const SeqNum msgSeqNum = msg->seqNumber();
   const Digest msgDigest = msg->digestOfState();
   const bool msgIsStable = msg->isStableState();
-
+  const EpochNum msgEpochNum = msg->epochNumber();
+  EpochNum replicasLastKnownEpochVal = 0;
   LOG_INFO(GL,
            KVLOG(msg->senderId(), msgGenReplicaId, msgSeqNum, msg->size(), msgIsStable)
                << ", digest: " << msgDigest.toString());
-
+  // Reconfiguration cmd block is synced to RO replica via reserved pages
+  auto epochNumberFromResPages = ReconfigurationCmd::instance().getReconfigurationCommandEpochNumber();
+  if (epochNumberFromResPages.has_value()) {
+    replicasLastKnownEpochVal = epochNumberFromResPages.value();
+  }
   // not relevant
   if (!msgIsStable || msgSeqNum <= lastExecutedSeqNum) return;
 
   // previous CheckpointMsg from the same sender
   auto pos = tableOfStableCheckpoints.find(msgGenReplicaId);
-  if (pos != tableOfStableCheckpoints.end() && pos->second->seqNumber() >= msgSeqNum) return;
+  if (pos != tableOfStableCheckpoints.end() &&
+      (pos->second->seqNumber() >= msgSeqNum || msgEpochNum < replicasLastKnownEpochVal))
+    return;
   if (pos != tableOfStableCheckpoints.end()) delete pos->second;
   CheckpointMsg *x = new CheckpointMsg(msgGenReplicaId, msgSeqNum, msgDigest, msgIsStable);
+  x->setEpochNumber(msgEpochNum);
```

Comment:
This is to handle the case where msgEpochNumber > replicaLastKnownEpochVal?

Reply:
yes